web/filament-js/extensions.js

```diff
@@ -185,9 +185,9 @@ Filament.loadClassExtensions = function() {
     /// engine ::argument:: [Engine]
     /// bufferIndex ::argument:: non-negative integer
     /// buffer ::argument:: asset string, or Uint8Array, or [Buffer]
-    Filament.VertexBuffer.prototype.setBufferAt = function(engine, bufferIndex, buffer) {
+    Filament.VertexBuffer.prototype.setBufferAt = function(engine, bufferIndex, buffer, byteOffset = 0) {
```

Comment:
To clarify:  Does this assume, then, that the ArrayBuffer being passed in and the existing VertexBuffer essentially mirror one another in terms of size and layout?  Meaning, the ArrayBuffer is kind of getting overlaid onto the VertexBuffer?

If so (unless I misunderstand), I believe our usage case would still require an offset parameter because we are setting up multiple geometries into a single VertexBuffer.  Each geometry creates its own "source" buffer, which we append onto the VertexBuffer.  The "source" buffers are not shared (non-contiguous) between the geometries creating them.  I hope that makes sense.

Reply:
I understand your confusion.  In Filament, a "VertexBuffer" is actually more like a traditional vertex array.  IMO we named it incorrectly.

A single "VertexBuffer" actually contains 1 or more low-level buffers, and each of these can be partitioned into 1 or more attributes in any way you need.  In JavaScript, each of these low-level buffers is populated from a typed array.

So, we do not need offsets for the `setBuffer` and `setBufferAt` API's.  However the thing that *is* missing (which you partially fixed) are the `setGeometry` capabilities within RenderableManager.  Specifically, these overloads are missing from the current JS API:

```
Builder& geometry(size_t index, PrimitiveType type, VertexBuffer* vertices, IndexBuffer* indices, size_t offset, size_t minIndex, size_t maxIndex, size_t count) noexcept;
Builder& geometry(size_t index, PrimitiveType type, VertexBuffer* vertices, IndexBuffer* indices, size_t offset, size_t count) noexcept; //!< \overload
```

In JS, maybe we could call the first one  `geometryMinMax(...)`.  You chose `geometryOffset` for the second one, which is fine.

Reply:
Still a bit confused.  Going back to the original question about offset:
We have chosen to use only one buffer per vertex buffer and interleave all of the attributes (pos, uv, tangents) into that buffer.
Let's say we have two "sub-objects" that are both using the same vertex buffer.  SubObject A has 10 verts, SubObject B has 5.
We create a vertex buffer with room for 15 vertices.
SubObject A creates a source buffer with its 10 vertices, and sets them into the vertex buffer at offset 0.
SubObject B creates a new source buffer with 5 vertices, and calls VertexBuffer.setBufferAt() with an byteOffset of 10 (ignoring actual vertex byte-size, for sake of discussion).

In JavaScript, how do we use the TypedArray offset to accomplish this?

In C++, this is what we're doing:
```
// SubObject A
vertexBuffer->setBufferAt(engine, 0, buffDescriptor1, 0); // Add 10 verts at offset 0
// SubObjectB
vertexBuffer->setBufferAt(engine, 0, buffDescriptor2, 10); // Add 5 verts at offset 10
```

Reply:
I'm still a little surprised that this argument is necessary since `setBuffer` consumes a typed array, which is a data view that already contain [offset](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/byteOffset).  i.e. `setBuffer(engine, myUint8Array, 10)` is equivalent to `setBuffer(engine, new Uint8Array(myUint8Array.buffer, myUint8Array.byteOffset + 10))`.

Anyway: let's just merge this though since it will definitely improve our parity with C++, and it sounds like it will help you.

Reply:
Great!  Thanks.  We've got the comments and the geometryMinMax added, so it should be good to go, unless you see any other issues you would like addressed.